src/upload-button.tsx

```diff
@@ -0,0 +1,351 @@
+                    continue;
+                }
+
+                if (resolution.applyToAll && resolution.replace)
+                    replaceAll = true;
+
+                toUploadFiles.push(inputFile);
+            } else {
+                toUploadFiles.push(inputFile);
+            }
+        }
+
+        if (toUploadFiles.length === 0) {
+            resetInput();
+            return;
+        }
+
+        window.addEventListener("beforeunload", beforeUnloadHandler);
+
+        const cancelledUploads = [];
+        await Promise.allSettled(toUploadFiles.map(async (file: File) => {
+            const tmp_path = path.slice();
+            tmp_path.push(file.name);
+            const destination = tmp_path.join('/');
+            const abort = new AbortController();
+
+            setUploadedFiles(oldFiles => {
+                return {
+                    [file.name]: { file, progress: 0, cancel: () => abort.abort() },
+                    ...oldFiles,
+                };
+            });
+
+            try {
+                await upload(destination, file, (progress) => {
+                    const now = performance.now();
+                    if (now < next_progress)
+                        return;
+                    next_progress = now + 200; // only rerender every 200ms
+                    setUploadedFiles(oldFiles => {
+                        const oldFile = oldFiles[file.name];
+                        return {
+                            ...oldFiles,
+                            [file.name]: { ...oldFile, progress },
+                        };
+                    });
+                }, abort.signal);
+                // TODO: pass { superuser: try } depending on directory owner
+            } catch (exc) {
+                cockpit.assert(exc instanceof Error, "Unknown exception type");
+                if (exc instanceof DOMException && exc.name === 'AbortError') {
+                    addAlert(_("Cancelled"), AlertVariant.warning, "upload",
+                             cockpit.format(_("Cancelled upload of $0"), file.name));
+                } else {
+                    addAlert(_("Upload error"), AlertVariant.danger, "upload-error", exc.toString());
+                }
+                cancelledUploads.push(file);
+            } finally {
+                setUploadedFiles(oldFiles => {
+                    const copy = { ...oldFiles };
+                    delete copy[file.name];
+                    return copy;
+                });
+            }
+        }));
+
+        resetInput();
+        window.removeEventListener("beforeunload", beforeUnloadHandler);
+
+        // If all uploads are cancelled, don't show an alert
+        if (cancelledUploads.length !== toUploadFiles.length) {
+            addAlert(_("Upload complete"), AlertVariant.success, "upload-success", _("Successfully uploaded file(s)"));
+        }
+    };
+
+    const isUploading = Object.keys(uploadedFiles).length !== 0;
```

Comment:
I don't understand this section.  Aren't we here because all of the uploads finished (`await allSettled`)?  Or is this in case duplicate copies of this function are running?  (ie: separately selected sets of files?)  If that's the case, the handling here seems a bit inconsistent....

Reply:
(This is the only thing outstanding for a +1 from me.)

Reply:
This is outside of the `onUpload` function but in the `render()` function

Reply:
Ah right!  I missed the out-dent.  Thanks!